packages/ra-core/src/dataProvider/useDeleteMany.ts

```diff
@@ -210,101 +215,85 @@ export const useDeleteMany = <
+            variables: Partial<UseDeleteManyMutateParams<RecordType>> = {},
+            context: unknown
+        ) => {
+            if (mode.current === 'pessimistic') {
+                // update the getOne and getList query cache with the new result
+                const {
+                    resource: callTimeResource = resource,
+                    ids: callTimeIds = ids,
+                } = variables;
+                updateCache({
+                    resource: callTimeResource,
+                    ids: callTimeIds,
+                });
 
-                    if (reactMutationOptions.onSuccess) {
-                        reactMutationOptions.onSuccess(
-                            data,
-                            variables,
-                            context
-                        );
-                    }
-                    // call-time success callback is executed by react-query
-                }
-            },
-            onSettled: (
-                data: RecordType['id'][],
-                error: MutationError,
-                variables: Partial<UseDeleteManyMutateParams<RecordType>> = {},
-                context: { snapshot: Snapshot }
-            ) => {
-                if (
-                    mode.current === 'optimistic' ||
-                    mode.current === 'undoable'
-                ) {
-                    // Always refetch after error or success:
-                    context.snapshot.forEach(([key]) => {
-                        queryClient.invalidateQueries(key);
-                    });
+                if (reactMutationOptions.onSuccess) {
+                    reactMutationOptions.onSuccess(data, variables, context);
                 }
+                // call-time success callback is executed by react-query
+            }
+        },
+        onSettled: (
+            data: RecordType['id'][],
+            error: MutationError,
+            variables: Partial<UseDeleteManyMutateParams<RecordType>> = {},
+            context: { snapshot: Snapshot }
+        ) => {
+            if (mode.current === 'optimistic' || mode.current === 'undoable') {
+                // Always refetch after error or success:
+                context.snapshot.forEach(([queryKey]) => {
+                    queryClient.invalidateQueries({ queryKey });
+                });
+            }
 
-                if (reactMutationOptions.onSettled) {
-                    return reactMutationOptions.onSettled(
-                        data,
-                        error,
-                        variables,
-                        context
-                    );
-                }
-            },
-        }
-    );
+            if (reactMutationOptions.onSettled) {
+                return reactMutationOptions.onSettled(
+                    data,
+                    error,
+                    variables,
+                    context
+                );
+            }
+        },
+    });
 
     const mutate = async (
```

Comment:
you need to add calltime/runtime side effect discrimination

Reply:
Added